src/mcpdotnet/Client/McpClientExtensions.cs

```diff
@@ -264,6 +268,143 @@ public static Task<CallToolResponse> CallToolAsync(this IMcpClient client, strin
             cancellationToken);
     }
 
+    /// <summary>Gets <see cref="AIFunction"/> instances for all of the tools available through the specified <see cref="IMcpClient"/>.</summary>
+    /// <param name="client">The client for which <see cref="AIFunction"/> instances should be created.</param>
+    /// <param name="cancellationToken">A token to cancel the operation.</param>
+    /// <returns>A task containing a list of the available functions.</returns>
+    public static async Task<IList<AIFunction>> GetAIFunctionsAsync(this IMcpClient client, CancellationToken cancellationToken = default)
+    {
+        Throw.IfNull(client);
+
+        List<AIFunction> functions = [];
+        await foreach (var tool in client.ListToolsAsync(cancellationToken).ConfigureAwait(false))
+        {
+            functions.Add(AsAIFunction(client, tool));
+        }
+
+        return functions;
+    }
+
+    /// <summary>Gets an <see cref="AIFunction"/> for invoking <see cref="Tool"/> via this <see cref="IMcpClient"/>.</summary>
+    /// <param name="client">The client with which to perform the invocation.</param>
+    /// <param name="tool">The tool to be invoked.</param>
+    /// <returns>An <see cref="AIFunction"/> for performing the call.</returns>
+    /// <remarks>
+    /// This operation does not validate that <paramref name="tool"/> is valid for the specified <paramref name="client"/>.
+    /// If the tool is not valid for the client, it will fail when invoked.
+    /// </remarks>
+    public static AIFunction AsAIFunction(this IMcpClient client, Tool tool)
+    {
+        Throw.IfNull(client);
+        Throw.IfNull(tool);
+
+        return new McpAIFunction(client, tool);
+    }
+
+    /// <summary>
+    /// Converts the contents of a <see cref="CreateMessageRequestParams"/> into a pair of
+    /// <see cref="IEnumerable{ChatMessage}"/> and <see cref="ChatOptions"/> instances to use
+    /// as inputs into a <see cref="IChatClient"/> operation.
+    /// </summary>
+    /// <param name="requestParams"></param>
+    /// <returns>The created pair of messages and options.</returns>
+    internal static (IEnumerable<ChatMessage> Messages, ChatOptions? Options) ToChatClientArguments( // TODO: Should this be exposed publicly?
```

Comment:
Why not IList<ChatMessage>?

Reply:
Just because Get{Streaming}ResponseAsync accept an IEnumerable, so I was keeping it consistent. But I could change this to be the more specific interface.